### DISCUSSION TITLE
Update not create new dto

### DISCUSSION THREAD

**Discussion by robert-abram** [spatie/laravel-data#97] [Q&A]
Helo,
In my application i'v created Data class and also use this class as cast

```php
class JsonTestData extends Data
{
    public function __construct(
        public ?string $type = null,
        public ?Carbon $date = null,
    ) {
    }
}
```

To update this model i use modal with only one field to update property date. When i construct new data class from request i loose my type property that was previously saved in db. How i can merge partial update to my Data class. NOw this is my code to do this.

```php
$DataRequest = JsonTestData::from($request);

$model = new Model();
$model->json_column = collect($$model->json_column)->merge(collect($DataRequest)->filter())
```

**Comment by ragulka** [comment]
I think I have a similar use-case, where I need to update some properties of a data object after it has been created, while also ensuring these properties will be cast to correct types and possibly validated. It doesn't look like there are any setters or update methods, though. 

The only workaround at the moment I can think of is to convert the existing data object to an array, merge the new values to the array and then create a new data object.

```
$data = DataObject::from( $request );
$data = DataObject::from( array_merge( $data->toArray(), ['foo' => '12' ] ) );
```

It would be nicer if I could just do:
```
$data = DataObject::from( $request );
$data->updateWith( ['foo' => '12'] );
```

**Comment by rubenvanassche** [comment]
Since the data object is just a plain PHP object, why not do the following?

```php
$data = DataObject::from($request);
$data->foo = 12;
```

**Reply by ragulka** [reply]
It bypasses type casting & data validation

**Comment by Nir-An** [comment]
> Helo, In my application i'v created Data class and also use this class as cast
> 
> ```
> class JsonTestData extends Data
> {
>     public function __construct(
>         public ?string $type = null,
>         public ?Carbon $date = null,
>     ) {
>     }
> }
> ```
> 
> To update this model i use modal with only one field to update property date. When i construct new data class from request i loose my type property that was previously saved in db. How i can merge partial update to my Data class. NOw this is my code to do this.
> 
> ```
> $DataRequest = JsonTestData::from($request);
> 
> $model = new Model();
> $model->json_column = collect($$model->json_column)->merge(collect($DataRequest)->filter())
> ```
Hi @abrampl and @ragulka 
I think for situations like this, you can use [fromArray function](https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/creating-a-data-object#content-magical-creation) and send the final array after you merge the collections.
or use fromMultiple method

```php
    public static function fromMultiple(Model $model, DataRequest $dataRequest): self
    {
       $someNewData = collect($$model->json_column)->merge(collect($DataRequest)->filter())
        return new self(
          ....
        );
    }
```

```
Instead of the default method, the fromModel method will be called to create a data object from the found model.

You're truly free to add as many from methods as you want. For example, you could add one to create a data object from a string:
```

**Comment by rubenvanassche** [comment]
As @Nir-An suggets,

I think it is better to create a new data object, by using a method which combines the arguments in the creating process it is even easier.

By definition, data transfer objects should be immutable, so actually you shouldn't change a DTO once it is created. When you want to change a DTO, then you should create a new one. Which can avoid a lot of problems later on when you're using the DTO's.

Now we're not enforcing such a pattern in this package, but it is something to think about.

As for the setters, I'm not completely opposed to it so I'm always open for a PR implementing this.
